main.go

```diff
@@ -0,0 +1,177 @@
+func webSocketHandler(w http.ResponseWriter, r *http.Request) {
+	//Upgrade this connection to a WebSocket connection
+	ws, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		log.Println(err)
+	}
+
+	// Create a client and assign it a Unique ID
+	client := Client{
+		Id:         autoId(),
+		Connection: ws,
+	}
+
+	// Send a message to the client
+	fmt.Printf("Client Connected:%s", client.Id)
+	err = ws.WriteMessage(1, []byte("Hi Client!"))
+	if err != nil {
+		log.Println(err)
+	}
+
+	// Add client to the list of clients
+	ps.AddClient(client)
+
+	// Listen indefinitely for new messages coming through on our WebSocket connection
+	for {
+		// Read a message
+		messageType, p, err := ws.ReadMessage()
+		if err != nil {
+			log.Println(err)
+			return
+		}
+		// Print out the message for clarity
+		log.Println(string(p))
+
+		// Send a message indicating the message was received
+		response := []byte("Server received the message!")
+		if err := ws.WriteMessage(messageType, response); err != nil {
+			log.Println(err)
+			return
+		}
+
+		// Call the handler to handle the received message from the client
+		ps.HandleRecvdMessage(client, messageType, p)
+	}
+
+}
+
+// Function to configure and handle the HTTP routes for the server.
+// It sets up two routes: one for serving static files and another for handling
+// WebSocket connections. The static route serves files from the "static" directory
+// and the WebSocket route uses the webSocketHandler function to handle incoming
+// WebSocket connections. 
+func setupRoutes() {
+	// Serve static files from the static directory
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		http.ServeFile(w, r, "static")
+	})
+	// Handle WebSocket connections using the webSocketHandler function
+	http.HandleFunc("/ws", webSocketHandler)
+}
+
+func main() {
+	fmt.Println("This is the main function of the Server")
+	setupRoutes()
+        // The server will listen on port 8080
+	if err := http.ListenAndServe(":8080", nil); err != nil {
+		log.Fatal(err)
+	}
+}
+
+// Function to add a new client to the list
+// Parameters:
+// client: Client - The client to be added to the list.
+// Returns:
+// *PubSub - A pointer to the updated PubSub instance after adding the client.
+func (ps *PubSub) AddClient(client Client) *PubSub {
+	ps.Clients = append(ps.Clients, client)
+	fmt.Println("Adding new client to the list", client.Id, len(ps.Clients))
+	payload := []byte("Hello Client ID" + client.Id)
+	client.Connection.WriteMessage(1, payload)
+	return ps
+}
+
+// Function to remove a client from the list
+// Parameters:
+// client: Client - The client to be removed from the list.
+// Returns:
+// *PubSub - A pointer to the updated PubSub instance after removing the client.
+func (ps *PubSub) RemoveClient(client Client) *PubSub {
+	for i, cl := range ps.Clients {
+		if cl.Id == client.Id {
+			ps.Clients = append(ps.Clients[:i], ps.Clients[i+1:]...)
+		}
+	}
```

Comment:
What data structure could you use for `O(1)` lookup time instead of the current `O(N)`?